lib/util/progress.coffee

```diff
@@ -1,53 +1,204 @@
-# TODO: move determinate progress bars to the front
-# TODO: hover UI with progress stack, descriptions
+{Emitter}  = require 'atom'
+inkTooltip = require './tooltip'
+
+# Progress Bars
+#
+# This module provides an API for displaying progress bars in Atom. The methods
+# below allow modifing the stack of progress bars, which is represented by
+# corresponding UI elements:
+#    - A "global" progress bar is shown in the status bar. In pratice, this is
+#      the first determinate progress bar in the stack. If there is none, the
+#      first element (which then is indeterminate) will be shown instead. If the
+#      stack is empty, an empty progress bar will be shown instead.
+#    - Hovering over that status bar element will show the complete stack as
+#      an overlay.
+#    - Hovering over the actual progress bar of one such stack element will show
+#      the message correesponding to that progress bar, if there is any.
+#
+# Methods:
+#
+# add(p = {progress: 0})
+#    Create and return a ProgressBar with the initial properties specified by `p`,
+#    which has the following methods available to it:
+#
+#    p.setProgress(prog)
+#        Updates `p`s progress. `prog` is a number between -∞ and 1; if it is
+#        negative, a indeterminate progress bar will be displayed.
+#
+#    p.setLeftText(t), p.setRightText(t), p.setMessage(t)
+#        Sets the text displayed to the left of the progress bar, right of the
+#        progress bar, or when hovering over it.
+#
+#    p.destroy()
+#        Destroys `p` and removes it from the display stack.
 
 module.exports =
+  stack: []
 
-  deactivate: ->
-    @tile?.destroy()
+  activate: ->
+    return if @activated
+    @activated = true
+    @emitter = new Emitter
 
-  metres: []
+    @overlay = @stackView()
+    @view = @tileView()
 
-  push: (p = {}) ->
-    if @metres.indexOf(p) == -1
-      @metres.push p
-      if @metres.length == 1
-        @showTile p
-      p.destroy = => @remove p
-      p
+    @tooltip = new inkTooltip @view, @overlay, => @hasDeterminateBars()
 
-  remove: (p) ->
-    i = @metres.indexOf p
-    @metres.splice i, 1 if i > -1
-    if @metres.length == 0
-      @showTile progress: 0
-    if i == 0 and @metres.length > 0
-      @showTile @metres[0]
+    @tile = @statusBar?.addLeftTile
+      item: @view
+      priority: -1
+
+  deactivate: ->
+    @tooltip?.destroy()
+    @tile?.destroy()
 
   consumeStatusBar: (bar) ->
     @statusBar = bar
 
+
+  create: (p = {progress: 0}) ->
+    @activate()
```

Comment:
`@activate` handles creating the statusbar element too, and so far we've only done that once the first client wants to display progress.